apps/Arm/vm_qemu_virtio/src/trace.c

```diff
@@ -37,24 +42,92 @@ unsigned int trace_index(const char *str)
     return trace_names_count - 1;
 }
 
-void trace_init(vm_t *vm)
+static inline int vka_cnode_copy(seL4_CPtr slot_src, seL4_CPtr slot_dst, seL4_CapRights_t rights)
 {
-    int error = vka_alloc_frame(&_vka, seL4_LargePageBits, &kernel_trace_frame);
-    if (error) {
-	    ZF_LOGF("Failed to allocate large page for kernel log");
-    }
-    error = seL4_BenchmarkSetLogBuffer(kernel_trace_frame.cptr);
+    cspacepath_t src_path, dest_path;
+
+    vka_cspace_make_path(&_vka, slot_src, &src_path);
+    vka_cspace_make_path(&_vka, slot_dst, &dest_path);
+
+    return seL4_CNode_Copy(
+               /* _service */      dest_path.root,
+               /* dest_index */    dest_path.capPtr,
+               /* destDepth */     dest_path.capDepth,
+               /* src_root */      src_path.root,
+               /* src_index */     src_path.capPtr,
+               /* src_depth */     src_path.capDepth,
+               /* rights */        rights
+           );
+}
+
+void trace_init_shared_mem(vm_t *vm, const char *name,
+		const uint64_t mem_addr, const uint64_t mem_size,
+		vka_object_t *local_mem, void **vm_memory, void **vmm_memory)
+{
+    /* virtual address to use on VM side,
+     * otherwise it will allocate @0x10000000
+     * and conflicts with VM's memory
+     */
+    void *vaddr = (void *)mem_addr;
+
+    /* Allocate memory  */
+    int error = vka_alloc_frame(&_vka, mem_size, local_mem);
     if (error) {
-	    ZF_LOGF("Cannot set kernel log buffer");
+	    ZF_LOGF("%s: failed to allocate pages for kernel log, size: 0x%lx", name, mem_size);
     }
-    kernel_trace = vspace_map_pages(&_vspace, &kernel_trace_frame.cptr, NULL, seL4_AllRights, 1, seL4_LargePageBits, true);
-    if (!kernel_trace) {
-        ZF_LOGF("Cannot map kernel log buffer to VMM vspace");
+    ZF_LOGE("%s: allocated page @0x%lx", name, local_mem->cptr);
+
+    /* Reserve VM virtual memoury 2MPage@vaddr */
```

Comment:
I wouldn't comment on this typo if the code itself wasn't so good.

Reply:
I checked this line 5 times.. or maybe even 10 -- sometimes I cannot see even such obvious things. Thanks.